gcovr/merging.py

```diff
@@ -240,15 +275,36 @@ def merge_function(
 
     If ``options.ignore_function_lineno`` is set,
     the two function coverage objects can have differing line numbers.
+    With following flags the merge mode can be defined:
+      - ``options.merge_function_use_line_zero``
+      - ``options.merge_function_use_line_min``
+      - ``options.merge_function_use_line_max``
+      - ``options.separate_function``
     """
     assert left.name == right.name
     if not options.ignore_function_lineno:
-        assert left.lineno == right.lineno
-
-    left.count += right.count
-
-    if options.ignore_function_lineno:
-        left.lineno = max(left.lineno, right.lineno)
+        assert left.count.keys() == right.count.keys()
+
+    for right_lineno in right.count.keys():
+        if right_lineno in left.count:
+            lineno = right_lineno
+        elif options.separate_function:
+            lineno = right_lineno
+            if lineno not in left.count:
+                left.count[lineno] = 0
+        else:
+            if options.merge_function_use_line_zero:
+                lineno = 0
+            elif options.merge_function_use_line_min:
+                lineno = min(*left.count.keys(), right_lineno)
+            elif options.merge_function_use_line_max:
+                lineno = max(*left.count.keys(), right_lineno)
+            # Set the sum at the desired line and clear all other data
+            count = sum(left.count.values())
+            left.count.clear()
+            left.count[lineno] = count
+
+        left.count[lineno] += right.count[right_lineno]
```

Comment:
Some of these options affect the `count` dicts as a whole, so it might be clearer to avoid the loops. For example (untested):

```suggestion
    # keep distinct counts for each line number
    if options.separate_function or options.ignore_function_lineno:
        for lineno, count in right.count.items():
            try:
                left.count[lineno] += count
            except KeyError:
                left.count[lineno] = count
        return left
        
    # merge all counts into an entry for a single line number
    if option.merge_function_use_line_zero:
        lineno = 0
    elif options.merge_function_use_line_min:
        lineno = min(*left.count.keys(), *right.count.keys())
    elif options.merge_function_line_use_max:
        lineno = max(*left.count.keys(), *right.count.keys())
    else:
        assert False, "unknown merge mode"
        
    left.count = { lineno: sum(left.count.values()) + sum(right.count.values()) }
```

Reply:
Good point.
Used this as a blueprint since it was incomplete.